### ISSUE TITLE
Recommend "command -v" over "if which X" or "if type X" to see if there's a command

### ISSUE THREAD

**Issue by david-a-wheeler** [koalaman/shellcheck#536]
Many people use "if which X" or "if type X" (or similar) to see if a command exists, but these aren't really portable.

The POSIX specification defines "command -v X", which is the portable way to do this - but a lot of people aren't aware of it.  Please point out the non-portable constructs, and explain that "command -v" is the portable way to determine it.

**Comment by Arthur2e5** [comment]
`type` is portable, at least XSI defines it: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/type.html

Using `command -v` suffers from the same problem as `type` over `which`, that it tries non-PATH things which would not be useful when you try to call other programs. `type -P`, a bashism, doesn't have the problem though.

**Comment by deryni** [comment]
Indeed. The situation with `which`, `type` and `command` is unfortunately complicated and difficult to navigate.

I find `command` especifically annoying since what it reports (with `-v`/`-V`) is very different then when it actually does (if used without `-v`).

**Comment by Arthur2e5** [comment]
For `command -v` there may still be a way through, basically:

```sh
whichable()( unalias -a; case "$(command -v "$1")" in (alias*) false;; (*/*) true;; (*) false;; esac )
```

**Comment by deryni** [comment]
`command` is a little better behaved, in that sense, with aliases then with functions.

    $ t() { echo t; }
    $ command -v t
    t
    $ command -V t
    t is a function
    t ()
    {
        echo
    }

Honestly, I'm not sure why `command -v/-V` reports aliases or functions *at all*.

It is a scope confusion that is largely the root of my problem with it.

**Comment by Arthur2e5** [comment]
uh-oh, I missed functions… Since there isn't a way to unset all functions, a plain-shell `which` like the one in [debianutils](https://anonscm.debian.org/cgit/users/clint/debianutils.git/tree/which) may be a better approach. Just make it a function in a subshell……

**Comment by deryni** [comment]
If you have bash then `type -P` and/or `type -pf` (after killing aliases) both work if you need to find the PATH-using value for the command.

That shell `which` is a nice tool though.

And while I'm fairly sure I've missed something. This command would seem to work to let you unset all functions.

    unset -f $(set | awk '/^[^[:space:]]* \(\) $/ {print $1}')

**Comment by koalaman** [comment]
@jonhiggs added a warning for `which` in e8fc094. Thanks!
